### PR TITLE
Completed Week 4 Assessment

### PR DESCRIPTION
**Pull request by scharlez**
# Submitting Week 4 Assessment

### ✅ Completed Files
Mark the following as confirmation that the files in this repository are complete.
- [ ✅ ] JavaScript Coding Challenges
- [ ✅ ] Ruby Coding Challenges
- [ ✅ ] Interview Questions

### 📝 Feedback
If you would like specific feedback on a particular problem please leave details here:
Please review the " ** NOTES TO INSTRUCTOR **: " Section

### CODE REVIEW

File: code-challenges.test.js [hunk 1]
```diff
@@ -15,35 +15,127 @@
...
+    return shuffleArray.shift() + shuffleArray.map(value => ({ value, sort: Math.random() })) + shuffleArray.sort((a, b) => a.sort - b.sort)
+    
+}
+
+//* RAN TEST (RESULTS BELOW): *//
+
+// (N/A - Nothing showed up when I ran the test, just said PASS but only as if I had the describe method aka Part A info only...it didn't register my logic.)
+
+// ** NOTES TO INSTRUCTOR **: I wasn't sure if I used the correct describe method setup based on the linked resource but I made my best guest and received a PASS for it. However, when it came to the second part of the question (the logic) as you can read above in the "RAN TEST" section it didn't register in Jest, it doesn't run the test for my logic for some reason. Any clarity on that? Regarding the logic, I wasn't sure if I should use Math.floor(Math.random() as I thought they may be mostly used for integers but I wasn't sure how else to randomize the elements of the array without it and my research wasn't providing many alternatives either. Side question should I have used .slice? 
+
```
Comment (triciasykes) [inline code-challenges.test.js:61]: Hi Steven - you were on the right track, just got a few things mixed up.  In the first part of your expect statement, you want to call your method and pass in the array ===> expect(firstShuffle(colors1)).   Then in the second part, you expect to get an array containing the remaining colors (after first element is removed) ===> (expect.arrayContaining(the expected outcome)).  final:        expect(shuffle(colors1)).toEqual(expect.arrayContaining(["yellow", "blue", "pink", "green"])). Hope this helps!

File: code-challenges.test.js [hunk 2]
```diff
@@ -15,35 +15,127 @@
...
+// Use .shift to remove the first element of original array.
+// Use .map to put each element in an object type setup and give them a random sort key. 
+// Use .sort to produce an a new arrangement using the random key. 
+// Return expected output. 
+
+const firstShuffle = (shuffleArray) => {
+    return shuffleArray.shift() + shuffleArray.map(value => ({ value, sort: Math.random() })) + shuffleArray.sort((a, b) => a.sort - b.sort)
+    
+}
+
```
Comment (triciasykes) [inline code-challenges.test.js:55]: Your thought process/logic here is solid. You are having a syntax/structure issue.  Don't use the `+` sign unless you are do a mathematical operation  or concatenating actual strings (not methods or variables).  If you start a new line at each `+`  and move the return to the last line, your code will work beautifully.

File: code-challenges.test.js [hunk 3]
```diff
@@ -15,35 +15,127 @@
...
+// Declare a function called minToMax 
+// Set the parameter as numericalArray.
+// Use .sort method to sort array from ascending order by specifying comparison function correctly. 
+// Return desired output. 
+
+// const minToMax= (numericalArray) => {
+//     numericalArray.sorts(function(a,b){return a-b;})
+//     return a - b; 
+// }
+
```
Comment (triciasykes) [inline code-challenges.test.js:96]: The method is actually .sort(), not .sorts().  Just a typo!

File: code-challenges.test.js [hunk 4]
```diff
@@ -15,35 +15,127 @@
...
+//* RAN TEST (RESULTS BELOW): *//
+
+// FAIL 
+// TypeError: numericalArray.sorts is not a function. 
+
+// ** NOTES TO INSTRUCTOR **: I was a bit surprised to see this error as a result of running my test, I thought my logic was...well logical. What did I do incorrectly? (Hence why my code is commented out here. )
```
Comment (triciasykes) [inline code-challenges.test.js:102]: I understand your logic - the structure is a little off.  I think you misread the question.  Your logic will put them in order, but it wants only the lowest and the highest number returned.

File: code-challenges.test.js [hunk 5]
```diff
@@ -15,35 +15,127 @@
...
+
+const arrayAlliance= (combinedArray) => {
+    combinedArray.concat() 
+    return combinedArray.filter(value) + combinedArray.indexOf(value)
+}
+//* RAN TEST (RESULTS BELOW): *//
+
+// ReferenceError: value is not defined
+
+// ** NOTES TO INSTRUCTOR **: I'm having some trouble understanding the difference between a Jest function and just writing out normal code in a console; for instance...in this problem, I know I didn't do the `.concat` correct because I didn't plug in or reference which arrays to combine and I know the error from the test I am getting is because I didn't previous define "value", but at the same time I'm not sure how to do so within the mechanisms of Jest, it's like I can write it out in a console and combine the arrays but I'm not sure how to get a passing test with the Jest parameters/above describe method information. Any insights or suggestions to get more clarity on this? I feel like I'm getting better but not really at the same time with these test, if any of this makes sense....
```
Comment (triciasykes) [inline code-challenges.test.js:141]: You are getting better!  Your test is solid!  You're thinking logically.  I think you're struggling with writing the function that passes the test.  I am happy to take some time to meet with you one on one to help you. Just let me know!

File: interview-questions.md [hunk 6]
```diff
@@ -7,52 +7,54 @@ Then, research the question to expand on your answer. Even if you feel you have
...
+2. PostgreSQL: a type of SQL that is open-source relational, highly stable database system that is mainly used for various web, mobile, analytics applications 
 
-3. Ruby on Rails:
+3. Ruby on Rails: is an open-source web applicaiton development framework from the Ruby language that simplies the creation of web applications
 
-4. ORM:
+4. ORM: stands for object-relational mapping, often used in PostgreSQL is a way of converting data between systems through OOP thus creating virtual object databases to do so. 
 
-5. Active Record:
+5. Active Record: help prevent one from having use SQL directly because it will perform queries on the databases for you  while also being compatible with most database systems like SQL.
```
Comment (triciasykes) [inline interview-questions.md:60]: Excellent work, as always!

File: ruby-challenges.rb [hunk 7]
```diff
@@ -14,20 +26,39 @@
...
 palindrome_tester1 = 'Racecar'
 # Expected output: 'Racecar is a palindrome'
 palindrome_tester2 = 'LEARN'
 # Expected output: 'LEARN is not a palindrome'
 palindrome_tester3 = 'Rotator'
 # Expected output: 'Rotator is a palindrome'
+
+p palindrome_string palindrome_tester1 
+p palindrome_string palindrome_tester2 
+p palindrome_string palindrome_tester3
```
Comment (triciasykes) [inline ruby-challenges.rb:64]: Awesome job!  Looks like you and Ruby are friends!  Excellent work!

**Review by triciasykes — APPROVED** [review]
Great work on this assessment, Steven!  I left comments on your questions.  Your tests and logic are good!  I think a good area to focus on is the structure of js functions.  I am happy to help you.  Again, your interview questions are stellar.  You are rocking Ruby!  Excellent work!

File: code-challenges.test.js (hunk 1 above)
Comment (scharlez) [reply code-challenges.test.js:61]: Thank you, that clears up some things!

File: code-challenges.test.js (hunk 3 above)
Comment (scharlez) [reply code-challenges.test.js:96]: Noted, thanks!

File: code-challenges.test.js (hunk 5 above)
Comment (scharlez) [reply code-challenges.test.js:141]: Absolutely! Thank you.

File: ruby-challenges.rb (hunk 7 above)
Comment (scharlez) [reply ruby-challenges.rb:64]: Yes, Ruby makes more logical sense to me at this point in my development than JS.

**Pull request merged**